src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1854 @@
+            // 3. The used vreg lives past the instruction.
+            // 4. The expected allocation of the vreg after the instruction is the clobber.
+            //
+            // Because of the way operand allocation works, edits to save and restore a vreg
+            // will have already been inserted during operand allocation if any of the following
+            // conditions are met:
+            // 1. The expected allocation afterwards is not a clobber.
+            // 2. There are multiple operands using the vreg with different constraints.
+            // 3. A def operand has the same clobber allocation assigned to it and
+            // the vreg lives past the instruction.
+            // Therefore, the presence of the vreg in `use_vregs_saved_and_restored`
+            // implies that it violates one of the conditions for the edits to be inserted.
+
+            let vreg = self.vreg_in_preg[clobbered_preg.index()];
+            if vreg != VReg::invalid() {
+                let vreg_isnt_mentioned_in_curr_inst = !self.vregs_in_curr_inst.contains(&vreg);
+                let vreg_lives_past_curr_inst = !self.vregs_first_seen_in_curr_inst.contains(&vreg);
+                if vreg_isnt_mentioned_in_curr_inst
+                    || (!self.use_vregs_saved_and_restored_in_curr_inst.contains(&vreg)
+                        && vreg_lives_past_curr_inst)
+                {
+                    trace!("Adding save and restore edits for {:?}", vreg);
+                    let preg_alloc = Allocation::reg(clobbered_preg);
+                    let slot = if self.vreg_spillslots[vreg.vreg()].is_valid() {
+                        self.vreg_spillslots[vreg.vreg()]
+                    } else {
+                        self.vreg_spillslots[vreg.vreg()] = self.allocstack(&vreg);
+                        self.vreg_spillslots[vreg.vreg()]
+                    };
+                    let slot_alloc = Allocation::stack(slot);
+                    self.add_move_later(
+                        inst,
+                        preg_alloc,
+                        slot_alloc,
+                        vreg.class(),
+                        InstPosition::Before,
+                        true
+                    );
+                    self.add_move_later(
+                        inst,
+                        slot_alloc,
+                        preg_alloc,
+                        vreg.class(),
+                        InstPosition::After,
+                        false,
+                    );
+                }
+            }
+        }
+        trace!("Done adding edits for clobbered registers");
+    }
+
+    /// If instruction `inst` is a branch in `block`,
+    /// this function places branch arguments in the spillslots
+    /// expected by the destination blocks.
+    /// 
+    /// The process used to do this is as follows:
+    /// 
+    /// 1. Move all branch arguments into corresponding temporary spillslots.
```

Comment:
This shouldn't be needed: because each vreg is assigned to a separate spillslot and each block parameter definition at the start of a block must be a distinct vreg (due to SSA), you're guaranteed to never have cycles. So you can just perform all the moves directly.

Reply:
It is possible to have cycles because the branch arguments need to be moved into the block parameter spillslots.
In the case where some or all of the branch arguments are a permutation of the block's parameters, like:
```
block1(v1, v2)
    1. goto block2

block2
    2. goto block1(v2, v1)
```
If the moves are inserted directly for block 2, then we'll have:

```
move from stack_v2 to stack_v1
move from stack_v1 to stack_v2
```
Which is incorrect.

Reply:
You are correct, it is actually possible to cause a cycle here.